File: server/wait_room.lua
local Server       = require "server"
local MatchManager = require "match_manager"
local Room         = require "room"
local log          = require "common.extra_libs.log"

local WaitRoom = {}

-- Map: Client -> Room Name
local clients = {}
-- Map: Room Name -> Room Obj
local rooms = {}

local function addClientToRoom(client, room)
    rooms[room] = rooms[room] or Room()
    clients[client] = room
    rooms[room]:addPlayer(client)
end

local function remClientFromRoom(client)
    local r = clients[client]
    clients[client] = nil
    rooms[r]:remPlayer(client)
    if r ~= 'none' and rooms[r]:empty() then
        rooms[r] = nil
    end
end

local function sendData()
    local room_list = {}
    for name, room in pairs(rooms) do
        table.insert(room_list, room:getData(name))
    end
    for cl in pairs(clients) do
        cl:send('client data', room_list)
    end
end

function WaitRoom.init()
    -- default room for all players
    rooms.none = Room()

    Server.on('connect', function(data, client)
        log.info('Client connected --', client)
        addClientToRoom(client, 'none')
        sendData()
    end)

    local change_room_schema = 'string'
    Server.on('change room', function(room, client)
        if not Server.checkSchema(client, change_room_schema, room) then return end
        if clients[client] ~= room then
            log.trace('Client ', client , ' changing room to ', room)
            remClientFromRoom(client)
            addClientToRoom(client, room)
            sendData()
        end
    end)

    local ready_schema = {ready = 'boolean', archetype = 'string'}
    Server.on('ready', function(args, client)
        if not Server.checkSchema(client, ready_schema, args) then return end
        log.trace('Client', client, 'is ready', args.ready)
        log.trace('Client', client, 'is using archetype', args.archetype)
        local r = rooms[clients[client]]
        local all_ready = r:updatePlayer(client, args)
        if clients[client] ~= 'none' and all_ready and r:atLeastTwoPlayers() then
            rooms[clients[client]] = nil
            local cl_list = {}
            log.trace('Creating match for clients ')
            for client, info in pairs(r.players) do
                -- we don't need this information anymore
                info.ready = nil
                table.insert(cl_list, {
                    client = client,
                    info = info
                })
                clients[client] = nil
                log.trace(client)
            end
            log.trace('')
            MatchManager.startMatch(cl_list)
        end
        sendData()
    end)

    Server.on('disconnect', function(data, client)
        log.info('Client disconnected --', client)
        if clients[client] == nil then
            -- Client is no longer in wait room
        else
            remClientFromRoom(client)
            sendData()
        end
    end)
end

return WaitRoom

File: server/wait_room.ts
import * as Server from './server';
import * as MatchManager from './match_manager';
import { Client } from 'common/extra_libs/sock';

class Room {
    name: string;
    clients: Set<Client>;
    constructor(name: string) {
        this.name = name;
        this.clients = new Set();
    }
}

const clients = new Map<Client, Room>();
const rooms = new Map<String, Room>();

function addClientToRoom(client: Client, room: string): void {
    if (!rooms.has(room))
        rooms.set(room, new Room(room));
    const r = rooms.get(room)!;
    clients.set(client, r);
    r.clients.add(client);
}

function remClientFromRoom(client: Client): void {
    const r = clients.get(client)!;
    clients.delete(client);
    r.clients.delete(client);
    if (r.name !== 'default' && r.clients.size === 0) {
        rooms.delete(r.name);
    }
}

function sendClientList(): void {
    const cl_list: Array<[Client, Room]> = [...clients.entries()];
    const list: Array<string> = [];
    for (const [client, room] of cl_list)
        list[client.getConnectId()] = room.name;
    for (const [client] of cl_list)
        client.send('client list', list);
}


export function init(): void {
    rooms['default'] = new Room('default');
    Server.on('connect', (data: never, client: Client) => {
        print('Client connected --', client);
        io.flush();
        addClientToRoom(client, 'default');
        sendClientList();
    });

    Server.on('change room', (room: string, client: Client) => {
        if (clients.get(client)!.name !== room) {
            print('Client', client, 'changing room to', room)
            remClientFromRoom(client);
            addClientToRoom(client, room);
            if (room !== 'default' && rooms.get(room)!.clients.size == 2) {
                const cls: [Client, Client][] = [...rooms.get(room)!.clients.entries()];
                const a = cls[0][0];
                const b = cls[1][0];
                remClientFromRoom(a);
                remClientFromRoom(b);
                MatchManager.startMatch(a, b);
            }
            sendClientList();
        }
    });
}